### PR TITLE
Implement Undo/Redo naming and history

### PR DESCRIPTION
**Pull request by rpatters1**
Resolves: #24509 Resolves: #20573 

- Adds a name for every call to `UndoStack::beginMacro` and displays these names in the menu and palettes.
- Adds `Undo/Redo History` to the Edit Menu and a dialog box to allow cycling through the history. This dialog box copies the simple single-list approach used by Dorico rather than the more complex dialog box used by Finale. Anyone who knows how to use the Dorico undo history should know immediately how to use this.

This PR incorporates the changes in #24543. I made a slight modification in `UndoRedoModel::load` to remove code duplication. Please review that part carefully, @cbjeukendrup.

The PR replaces #24587, and that PR is now closed.

- [X] I signed the [CLA](https://musescore.org/en/cla)
- [X] The title of the PR describes the problem it addresses
- [X] Each commit's message describes its purpose and effects, and references the issue it resolves
- [X] If changes are extensive, there is a sequence of easily reviewable commits
- [X] The code in the PR follows [the coding rules](https://github.com/musescore/MuseScore/wiki/CodeGuidelines)
- [X] There are no unnecessary changes
- [X] The code compiles and runs on my machine, preferably after each commit individually
- [ ] I created a unit test or vtest to verify the changes I made (if applicable)

### CODE REVIEW

**Comment by cbjeukendrup** [comment]
Rebased, and squashed the commits, since most changes were in one big commit anyway and then there were a few follow-up commits that were not very interesting on their own.

**Comment by rpatters1** [comment]
Looks like it isn't building. I'll check later this afternoon when I get back to my dev box. I was planning to do this myself later after the other commit was merged (rebasing on the base branch), but thanks for doing it.

**Comment by cbjeukendrup** [comment]
I made some corrections to the names of the undoable actions; that was the most time-consuming part. Tomorrow I'll review the new dialog code.

**Comment by rpatters1** [comment]
Thank you. I figures someone who knows the code better that I do should go over those.

**Comment by rpatters1** [comment]
Thank you for taking this under your wing. I had already thought about splitting `UndoRedoModel` but I got side-tracked.

**Comment by cbjeukendrup** [comment]
Update: just pushed some slightly larger amendments, and there is still a bit more to do, but I think we should do that in future PRs.

Here is what I did:
- slight cleanup of the UndoStack class
- bit of optimisation in `NotationUndoStack::undoRedoToIdx`
- code cleanup in UndoHistoryDialog.qml: eliminated some useless Row/ColumnLayouts; and adjusted the spacing a little bit. 
- split UndoRedoModel into UndoRedoToolbarModel and UndoRedoHistoryModel; their functionality is sufficiently disjoint, and this makes it easier to refactor either of them in the future

It would be great if you could now review my changes (reviewing commit-by-commit is recommended), to be sure I didn't make mistakes. 

What's left to do:
- I think `NotationUndoStack::undoRedoToIdx` functions slightly weirdly: I would expect that if `stack->currentIndex() == idx`, then nothing happens. Instead, it currently redoes one command in that case. To prevent confusion in the future, that should perhaps be changed. 
- Instead of doing a lot of logic in QML, it would be more efficient to rewrite UndoRedoHistoryModel as a `QAbstractListModel`. I've made a start at this, but it's not fully done yet; I'll post my progress somewhere, so that you can potentially build on that. 
- The UI/UX may need some tweaks. I'll ask the design team to come up with something; here are some ideas:
    - It would be nice if the OK button reflected what is going to happen, i.e. "Undo 1 action" or "Redo 4 actions". 
    - Perhaps there should be a "current state" row, that is unelectable; that might make it clearer how this dialog works. Selecting anything above this row would redo it; anything below would undo.
    - If you select, say, the 5th row below the "current state" row, then it might be nice if not just that row, but all 5 rows below "current state" are highlighted, to indicate that those 5 actions will be undone.
    - There is a slight chance that we might want to go back to Finale's two-column design; personally I think that wasn't that bad at all. If the designers indeed want that, that would change things a bit, also with regard to the QAbstractListModel story. But we'll see.

As said, it may be better to do this in next PRs. So after you review my changes, I'll merge this PR, and create an issue about the things that are left to do. Does that sound okay?

File: src/appshell/view/appmenumodel.h
Comment (rpatters1) [inline src/appshell/view/appmenumodel.h:78]: For some reason, this line here bugs me because it sticks out so much. I get why it's there though, and though it matters not at all, I would probably move the declaration to the bottom to avoid disrupting the march of top-level menu-maker functions. This is totally your call.

**Comment by rpatters1** [comment]
I have been reviewing your changes commit-by-commit as you have put them in, as much out of curiosity as anything else. So far, nothing has jumped out at me. I just went through them again, as well as unit testing the Undo/Redo functionality in MuseScore, and I can't find anything the matter. (As we used to say back when I did this for money, all my tests were unsuccessful because I did not find anything wrong.)

As to your comments, I replicated the behavior of Dorico's undo history window as closely as I thought made sense. If you are on the last line, it toggles between undo & redo, which makes sense to me but maybe not to a non-techie user. Since I wrote the code I have learned that there are a lot of complaints about the Dorico undo history. Tbh, I was looking for a way not to have to do a lot of qml because it was my first attempt at it and I'm still working through my QRP for QML. (QRP = Quota of Rotten Programs.)

The Finale way is definitely better, but what it really needs to be useful is the marker where the file was last saved. That part, at least, is definitely phase X. FWIW: I really dislike the idea of an unselectable row. Unselectable rows are almost always user abuse. I believe my code currently prefers Undo but toggles when you get to the top or bottom. Finale's window does not have an unselectable row, but the two columns make it much clearer what will happen. That, at least, should probably be phase 2 at the latest. But before diving in, perhaps we should wait to hear from the design team.

**Comment by rpatters1** [comment]
I should add that your plan sounds good to me.

File: src/appshell/view/appmenumodel.h
Comment (cbjeukendrup) [reply src/appshell/view/appmenumodel.h]: Yeah, I didn't know very well what to do with it, but perhaps the bottom is the best place indeed.

**Review by cbjeukendrup — APPROVED** [review]
Approving this as a first step; remaining tasks have been logged in #25327.

**Comment by cbjeukendrup** [comment]
(For the record: not sending this to QA (quality assurance) yet, because the UI code will still be revised in #25327.)

**Pull request merged**
